src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -22,26 +22,75 @@
+--   function is defined as the partial application of another defined function 
+--   or constructor) and partial applications that occur in the arguments of 
+--   the function declaration's right-hand side. 
+-- 
+--   We perform regular η-conversions on partial applications in proper 
+--   subexpressions of a function declaration's right-hand side. 
+-- 
+--   However, on the top-level, we add the missing arguments to the left-hand
+--   and right-hand sides of the function rule explicitly, without a lambda
+--   abstraction. 
+--   This is an optimization that allows the compiler to avoid some unnecessary 
+--   monadic lifting.
+--
 --   = Specification
 --
 --   == Preconditions
 --
 --   The arity of all constructors and functions must be known (i.e., there
---   must be corresponding environment entries).
+--   must be corresponding environment entries) and all function declarations 
+--   must be type annotated.
 --
 --   == Translation
 --
---   On the right-hand sides of function declarations, all of the largest
---   sub-expressions of the form
+--   Assume that we have the following function declaration. 
+--
+--   > f e₁ … eₖ = g @α₁ … @αₚ e₁ … eₘ
+--
+--   where @g@ is the name of an @n@-ary constructor or function declaration
+--   and @m < n@. This declaration is then replaced by 
+--
+--   > f e₁ … eₖ x₍ₘ₊₁₎ … xₙ = g @α₁ … @αₚ e₁ … eₘ x₍ₘ₊₁₎ … xₙ
+--   
+--   where x₍ₘ₊₁₎ … xₙ are @n-m@ fresh variables. 
 --
---   > f @α₁ … @αₚ e₁ … eₘ
+--   If a function rule has several alternatives for the right-hand side
+--   through (possibly nested) if or case expressions, the number of added
+--   arguments is determined by the minimum number of missing arguments for an
+--   alternative on the right-hand side.
 --
---   where @f@ is the name of an @n@-ary constructor or function declaration
---   and @m < n@ is replaced by a lambda abstraction
+--   Additionally, on the right-hand sides of function declarations, all of the
+--   largest sub-expressions of the form
 --
---   > \x₍ₘ₊₁₎ … xₙ -> f @α₁ … @αₚ e₁ … eₘ x₍ₘ₊₁₎ … xₙ
+--   > h @α₁ … @αᵣ e₁ … eₗ
 --
---   where @x₍ₘ₊₁₎ … xₙ@ are @n-m@ fresh variables.
+--   where @h@ is the name of an @p@-ary constructor or function declaration
+--   and @l < p@ are replaced by a lambda abstraction
+--
+--   > \y₍ₗ₊₁₎ … yₚ -> f @α₁ … @αᵣ e₁ … eₘ y₍ₗ₊₁₎ … yₚ
+--
+--   where @y₍ₗ₊₁₎ … yₚ@ are @p-l@ fresh variables. 
+-- 
+--   Both types of η-conversion may also be applied to the same expression.
+--   For example, the function 
+--   
+--  > f :: τ₁ -> τ₂ -> τ₃ > τ 
+--  > f e₁ = case e₁ of {
+--  >           c₁ -> g; 
+--  >           c₂ -> h 
+--  >   }
+--
+--  where @c₁@ and @c₂@ are the constructors of @τ₁@, 
+--  @g@ is a binary function of type @τ₂ -> τ₃ > τ@, and
+--  @h@ is a unary function of type @τ₂ -> (τ₃ > τ)@, 
+--  will be converted to 
+-- 
+--  > f e₁ x = case e₁ of
+--  >           c₁ -> \y -> g x y
+--  >           c₂ -> h x 
```

Comment:
My [previous comment](https://github.com/FreeProving/free-compiler/pull/46#discussion_r440710789) about using IR notation in examples also applies to this code block. It may also be a good idea to write the type annotations in IR notation to demonstrate that `modifyTopLevel` manipulates the type annotations.

```haskell
f (e₁ :: τ₁) :: τ₂ -> τ₃ -> τ = case e₁ of {
    c₁ -> g;
    c₂ -> h 
  }

f (e₁ :: τ₁) (x :: τ₂) :: τ₃ -> τ = case e₁ of {
    c₁ -> \y -> g x y
    c₂ -> h x 
  }
```